File: fasten-core/bindings/go/fasten_store_core_test.go
// Tests for the Go binding (fasten_core.go).
//
// Requires libfasten_core at link time:
//
//	cd fasten/fasten-core && cargo build --release --features all
//	cd bindings/go
//	CGO_LDFLAGS="-L../../../../target/release" go test -v ./...
//
// PostgreSQL tests are skipped when FASTEN_TEST_POSTGRES_DSN is absent.
package fastenstore_test

import (
	"encoding/json"
	"os"
	"testing"

	fastenstore "."
)

// ── helpers ───────────────────────────────────────────────────────────────────

func makeRow(id, code string) map[string]any {
	return map[string]any{
		"wire_version":    "1",
		"id":              id,
		"origin_id":       id,
		"monotonic_seq":   1,
		"timestamp":       "2026-05-07T00:00:00.000Z",
		"code":            code,
		"action":          "test",
		"severity":        "info",
		"service_id":      "test-svc",
		"source_node_id":  "node-1",
		"actor":           "tester",
		"actor_kind":      "user",
		"target":          "res-1",
		"category":        "test",
		"domain":          "test",
		"method":          "sdk",
		"request_id":      "req-001",
		"detail":          map[string]any{"key": "value"},
	}
}

func pgDSN(t *testing.T) string {
	t.Helper()
	dsn := os.Getenv("FASTEN_TEST_POSTGRES_DSN")
	if dsn == "" {
		t.Skip("FASTEN_TEST_POSTGRES_DSN not set")
	}
	return dsn
}

// ── Version ───────────────────────────────────────────────────────────────────

func TestVersion(t *testing.T) {
	v := fastenstore.Version()
	if v == "" {
		t.Fatal("Version() returned empty string")
	}
	t.Logf("library version: %s", v)
}

// ── SQLite ────────────────────────────────────────────────────────────────────

func TestSQLite_OpenAndPing(t *testing.T) {
	store, err := fastenstore.Open("sqlite", ":memory:", "audit_log")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	if err := store.Ping(); err != nil {
		t.Fatalf("Ping: %v", err)
	}
}

func TestSQLite_Insert(t *testing.T) {
	store, err := fastenstore.Open("sqlite", ":memory:", "audit_log")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	if err := store.Insert(makeRow("evt-go-sqlite-001", "TEST")); err != nil {
		t.Fatalf("Insert: %v", err)
	}
}

func TestSQLite_InsertIdempotent(t *testing.T) {
	store, err := fastenstore.Open("sqlite", ":memory:", "audit_log")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	row := makeRow("evt-go-idem-001", "TEST")
	if err := store.Insert(row); err != nil {
		t.Fatalf("first insert: %v", err)
	}
	if err := store.Insert(row); err != nil {
		t.Fatalf("idempotent insert: %v", err) // INSERT OR IGNORE
	}
}

func TestSQLite_InsertJSON(t *testing.T) {
	store, err := fastenstore.Open("sqlite", ":memory:", "audit_log")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	b, _ := json.Marshal(makeRow("evt-go-json-001", "TEST"))
	if err := store.InsertJSON(string(b)); err != nil {
		t.Fatalf("InsertJSON: %v", err)
	}
}

func TestSQLite_NullableColumns(t *testing.T) {
	store, err := fastenstore.Open("sqlite", ":memory:", "audit_log")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	row := makeRow("evt-go-null-001", "TEST")
	row["tenant_id"] = "tenant-abc"
	row["shipped_at"] = "2026-05-07T01:00:00.000Z"
	if err := store.Insert(row); err != nil {
		t.Fatalf("Insert with nullable cols: %v", err)
	}
}

func TestSQLite_PiiInDetail(t *testing.T) {
	store, err := fastenstore.Open("sqlite", ":memory:", "audit_log")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	row := makeRow("evt-go-pii-001", "TEST")
	row["pii_in_detail"] = true
	if err := store.Insert(row); err != nil {
		t.Fatalf("Insert pii: %v", err)
	}
}

func TestSQLite_InvalidTableNameReturnsError(t *testing.T) {
	_, err := fastenstore.Open("sqlite", ":memory:", "bad-name!")
	if err == nil {
		t.Fatal("expected error for invalid table name, got nil")
	}
}

func TestSQLite_UnknownBackendReturnsError(t *testing.T) {
	_, err := fastenstore.Open("nope", ":memory:", "audit_log")
	if err == nil {
		t.Fatal("expected error for unknown backend, got nil")
	}
}

// ── PostgreSQL ────────────────────────────────────────────────────────────────

func TestPostgres_OpenAndPing(t *testing.T) {
	dsn := pgDSN(t)
	store, err := fastenstore.Open("postgres", dsn, "fasten_go_sc_test")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	if err := store.Ping(); err != nil {
		t.Fatalf("Ping: %v", err)
	}
}

func TestPostgres_InsertIdempotent(t *testing.T) {
	dsn := pgDSN(t)
	store, err := fastenstore.Open("postgres", dsn, "fasten_go_sc_test")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()

	row := makeRow("evt-go-pg-001", "TEST")
	if err := store.Insert(row); err != nil {
		t.Fatalf("first insert: %v", err)
	}
	if err := store.Insert(row); err != nil {
		t.Fatalf("idempotent insert: %v", err) // ON CONFLICT DO NOTHING
	}
}

func TestPostgres_SchemaQualified(t *testing.T) {
	dsn := pgDSN(t)
	store, err := fastenstore.Open("postgres", dsn, "fasten_go_sc_schema.audit_rows")
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer store.Close()
	if err := store.Insert(makeRow("evt-go-schema-001", "TEST")); err != nil {
		t.Fatalf("Insert: %v", err)
	}
}

File: fasten-core/bindings/node/fasten_store_core.test.mjs
/**
 * Tests for the Node.js binding (fasten_store_core.mjs).
 *
 * Requires libfasten_store_core to be built:
 *   cd fasten/store-core && cargo build --release --features all
 *
 * Run:
 *   node --test fasten_store_core.test.mjs
 *
 * Skip gracefully when the library is absent so CI passes without it.
 */

import { test, describe, before, after } from 'node:test';
import assert from 'node:assert/strict';

// ── Library availability guard ────────────────────────────────────────────────

let FastenStore, FastenStoreError, libAvailable;

try {
  ({ FastenStore, FastenStoreError } = await import('./fasten_store_core.mjs'));
  libAvailable = true;
} catch {
  libAvailable = false;
}

function skip(name, fn) {
  if (!libAvailable) {
    test(name, { skip: 'libfasten_store_core not built' }, fn);
  } else {
    test(name, fn);
  }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function makeRow(id, code = 'TEST') {
  return {
    wire_version: '1', id, origin_id: id,
    monotonic_seq: 1,
    timestamp: '2026-05-07T00:00:00.000Z',
    code, action: 'test', severity: 'info',
    service_id: 'test-svc', source_node_id: 'node-1',
    actor: 'tester', actor_kind: 'user',
    target: 'res-1', category: 'test', domain: 'test',
    method: 'sdk', request_id: 'req-001',
    detail: { key: 'value' },
  };
}

const PG_DSN = process.env.FASTEN_TEST_POSTGRES_DSN;

// ── Version ───────────────────────────────────────────────────────────────────

skip('version() returns a non-empty string', () => {
  const v = FastenStore.version();
  assert.ok(typeof v === 'string' && v.length > 0, `bad version: ${v}`);
});

// ── SQLite ────────────────────────────────────────────────────────────────────

skip('sqlite: open :memory: and ping', () => {
  const store = FastenStore.open('sqlite', ':memory:', 'audit_log');
  try { store.ping(); }
  finally { store.close(); }
});

skip('sqlite: insert a row', () => {
  const store = FastenStore.open('sqlite', ':memory:', 'audit_log');
  try { store.insert(makeRow('evt-node-sqlite-001')); }
  finally { store.close(); }
});

skip('sqlite: insert is idempotent', () => {
  const store = FastenStore.open('sqlite', ':memory:', 'audit_log');
  const row = makeRow('evt-node-idem-001');
  try {
    store.insert(row);
    store.insert(row); // INSERT OR IGNORE — no throw
  } finally { store.close(); }
});

skip('sqlite: insert from pre-serialised JSON string', () => {
  const store = FastenStore.open('sqlite', ':memory:', 'audit_log');
  try { store.insert(JSON.stringify(makeRow('evt-node-json-001'))); }
  finally { store.close(); }
});

skip('sqlite: nullable tenant_id + shipped_at', () => {
  const store = FastenStore.open('sqlite', ':memory:', 'audit_log');
  const row = { ...makeRow('evt-node-null-001'), tenant_id: 'tenant-abc',
                shipped_at: '2026-05-07T01:00:00.000Z' };
  try { store.insert(row); }
  finally { store.close(); }
});

skip('sqlite: pii_in_detail flag', () => {
  const store = FastenStore.open('sqlite', ':memory:', 'audit_log');
  const row = { ...makeRow('evt-node-pii-001'), pii_in_detail: true };
  try { store.insert(row); }
  finally { store.close(); }
});

skip('sqlite: Symbol.dispose closes the store', () => {
  { using store = FastenStore.open('sqlite', ':memory:', 'audit_log');
    store.insert(makeRow('evt-node-dispose-001')); }
  // store.close() was called by Symbol.dispose
});

skip('invalid table name throws FastenStoreError', () => {
  assert.throws(
    () => FastenStore.open('sqlite', ':memory:', 'bad-name!'),
    FastenStoreError,
  );
});

skip('unknown backend throws FastenStoreError', () => {
  assert.throws(
    () => FastenStore.open('nope', ':memory:', 'audit_log'),
    FastenStoreError,
  );
});

// ── PostgreSQL (skipped without DSN) ──────────────────────────────────────────

if (PG_DSN) {
  skip('postgres: open and ping', () => {
    const store = FastenStore.open('postgres', PG_DSN, 'fasten_node_sc_test');
    try { store.ping(); }
    finally { store.close(); }
  });

  skip('postgres: insert idempotent', () => {
    const store = FastenStore.open('postgres', PG_DSN, 'fasten_node_sc_test');
    const row = makeRow('evt-node-pg-001');
    try { store.insert(row); store.insert(row); }
    finally { store.close(); }
  });

  skip('postgres: schema-qualified table', () => {
    const store = FastenStore.open('postgres', PG_DSN, 'fasten_node_sc_schema.audit_rows');
    try { store.insert(makeRow('evt-node-schema-001')); }
    finally { store.close(); }
  });
}
